### PR TITLE
unique hashes for spc txs

### PR DESCRIPTION
**Pull request by longwinterday**

### CODE REVIEW

**Comment by liam-lai** [comment]
Hi, can you fetch the latest code from dev-upgrade please? Your test is failed.

**Comment by wjrjerome** [comment]
Could you please create relevant unit or integration tests to cover the changes you made?

**Comment by wgr523** [comment]
I think you use this line `time.Now().UnixNano() / int64(time.Millisecond) % txMatchGasLimit` to avoid the transaction contains the same gas limit. However, there is still a small chance of 1/txMatchGasLimit such that the gas limit is the same. Have you considered using `time.Now().UnixNano() / int64(time.Millisecond)`?

**Comment by wgr523** [comment]
By the way, is it feasible to use incremental nonce, just like normal transactions? If nonce is incremented (`nonce+=1`), then nonce will be different for each transaction.

**Comment by longwinterday** [comment]
> I think you use this line `time.Now().UnixNano() / int64(time.Millisecond) % txMatchGasLimit` to avoid the transaction contains the same gas limit. However, there is still a small chance of 1/txMatchGasLimit such that the gas limit is the same. Have you considered using `time.Now().UnixNano() / int64(time.Millisecond)`?

No. It is impossible to use Gas Limit larger when pool Gas Limit https://github.com/XinFinOrg/XDPoSChain/blob/be3e9ebe05a43754a76092053782b61d9960e7fc/core/tx_pool.go#L608

However, if it be based on case that blocks appears every 2 seconds and it looks so, it is possible to use `time.Now().UnixNano() / int64(time.Millisecond * 100) % txMatchGasLimit` it is about 1/40_000_000 every 46days a chance to have same dynamicTxMatchGasLimit and the nonce should be kept but actually I can't find cases with triple and more not unique transactions. So the chances are about 1 not unique transaction every 105 000 years. 

Nonce can be also incremented but it won't guarantee that every block will have special transaction cause of possible errors. Like get address from in that case https://github.com/XinFinOrg/XDPoSChain/blob/22479f11ff3063b5d2d8468c356dc59a79570af9/core/state_processor.go#L452

**Comment by longwinterday** [comment]
> Could you please create relevant unit or integration tests to cover the changes you made?

Hmm. I do not see special test cases for these changes. I suppose all current test cover the worker's functional.

File: miner/worker.go [hunk 1]
```diff
@@ -701,7 +701,8 @@ func (self *worker) commitNewWork() {
 						return
 					}
 					nonce := work.state.GetNonce(self.coinbase)
-					tx := types.NewTransaction(nonce, common.HexToAddress(common.XDCXAddr), big.NewInt(0), txMatchGasLimit, big.NewInt(0), txMatchBytes)
+					dynamicTxMatchGasLimit := uint64(time.Now().UnixNano() / int64(time.Millisecond) % txMatchGasLimit)
+					tx := types.NewTransaction(nonce, common.HexToAddress(common.XDCXAddr), big.NewInt(0), dynamicTxMatchGasLimit, big.NewInt(0), txMatchBytes)
```
Comment (liam-lai) [inline miner/worker.go:705]: can you please create a function in types that avoid duplicate code all over the place

**Comment by wjrjerome** [comment]
> > Could you please create relevant unit or integration tests to cover the changes you made?
> 
> Hmm. I do not see special test cases for these changes. I suppose all current test cover the worker's functional.

Ideally, it would be best to simulate the issue in unit test prior to your fix. i.e it shall fail. 
Then run the test against your fix to verify it will pass.

**Review by wjrjerome — APPROVED** [review]

**Review by wgr523 — APPROVED** [review]

File: miner/worker.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (liam-lai) [reply miner/worker.go:705]: @longwinterday could you reply above please

**Comment by liam-lai** [comment]
@longwinterday will this change create a hard forking?

**Comment by gzliudan** [comment]
solved by #430 #501 #521

**Pull request closed without merging**